transports/websocket/src/framed.rs

```diff
@@ -0,0 +1,431 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use bytes::BytesMut;
+use crate::{error::Error, tls};
+use futures::{future::{self, Either, Loop}, prelude::*, try_ready};
+use libp2p_core::{
+    Transport,
+    either::EitherOutput,
+    multiaddr::{Protocol, Multiaddr},
+    transport::{ListenerEvent, TransportError}
+};
+use log::{debug, trace};
+use tokio_rustls::{client, server};
+use soketto::{base, connection::{Connection, Mode}, handshake::{self, Redirect, Response}};
+use std::io;
+use tokio_codec::{Framed, FramedParts};
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_rustls::webpki;
+use url::Url;
+
+/// Max. number of payload bytes of a single frame.
+const MAX_DATA_SIZE: u64 = 256 * 1024 * 1024;
+
+/// A Websocket transport whose output type is a [`Stream`] and [`Sink`] of
+/// frame payloads which does not implement [`AsyncRead`] or
+/// [`AsyncWrite`]. See [`crate::WsConfig`] if you require the latter.
+#[derive(Debug, Clone)]
+pub struct WsConfig<T> {
+    transport: T,
+    max_data_size: u64,
+    tls_config: tls::Config,
+    max_redirects: u8
+}
+
+impl<T> WsConfig<T> {
+    /// Create a new websocket transport based on another transport.
+    pub fn new(transport: T) -> Self {
+        WsConfig {
+            transport,
+            max_data_size: MAX_DATA_SIZE,
+            tls_config: tls::Config::client(),
+            max_redirects: 0
+        }
+    }
+
+    /// Return the configured maximum number of redirects.
+    pub fn max_redirects(&self) -> u8 {
+        self.max_redirects
+    }
+
+    /// Set max. number of redirects to follow.
+    pub fn set_max_redirects(&mut self, max: u8) -> &mut Self {
+        self.max_redirects = max;
+        self
+    }
+
+    /// Get the max. frame data size we support.
+    pub fn max_data_size(&self) -> u64 {
+        self.max_data_size
+    }
+
+    /// Set the max. frame data size we support.
+    pub fn set_max_data_size(&mut self, size: u64) -> &mut Self {
+        self.max_data_size = size;
+        self
+    }
+
+    /// Set the TLS configuration if TLS support is desired.
+    pub fn set_tls_config(&mut self, c: tls::Config) -> &mut Self {
+        self.tls_config = c;
+        self
+    }
+}
+
+impl<T> Transport for WsConfig<T>
+where
+    T: Transport + Send + Clone + 'static,
+    T::Error: Send + 'static,
+    T::Dial: Send + 'static,
+    T::Listener: Send + 'static,
+    T::ListenerUpgrade: Send + 'static,
+    T::Output: AsyncRead + AsyncWrite + Send + 'static
+{
+    type Output = BytesConnection<T::Output>;
+    type Error = Error<T::Error>;
+    type Listener = Box<dyn Stream<Item = ListenerEvent<Self::ListenerUpgrade>, Error = Self::Error> + Send>;
```

Comment:
Boxing goes in the opposite direction of #369.

Reply:
I would not say it is the "opposite direction", but that it does not improve w.r.t. the boxing situation in the websocket transport, considering that it [uses Boxes here](https://github.com/libp2p/rust-libp2p/blob/6a2a401458529d79a5398f770711235ac6bddb11/transports/websocket/src/desktop.rs#L74-L78) as well.